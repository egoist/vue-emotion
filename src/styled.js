import { h } from 'vue'
import { getRegisteredStyles, insertStyles } from '@emotion/utils'
import { serializeStyles } from '@emotion/serialize'

const ILLEGAL_ESCAPE_SEQUENCE_ERROR =
  process.env.NODE_ENV === 'production' ?
    '' :
    `You have illegal escape sequence in your template literal, most likely inside content's property value.
Because you write your CSS inside a JavaScript string you actually have to do double escaping, so for example "content: '\\00d7';" should become "content: '\\\\00d7';".
You can read more about this here:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#ES2018_revision_of_illegal_escape_sequences`

const createStyled = (tag, options = {}) => {
  if (process.env.NODE_ENV !== 'production') {
    if (tag === undefined) {
      throw new Error(
        'You are trying to create a styled element with an undefined component.\nYou may have forgotten to import it.'
      )
    }
  }

  const identifierName = options.label
  const targetClassName = options.target

  const isReal = tag.__emotion_real === tag
  const baseTag = (isReal && tag.__emotion_base) || tag

  return function (...args) {
    const styles =
      isReal && tag.__emotion_styles !== undefined ?
        tag.__emotion_styles.slice(0) :
        []

    if (identifierName !== undefined) {
      styles.push(`label:${identifierName};`)
    }

    if (args[0] === null || args[0].raw === undefined) {
      styles.push(...args)
    } else {
      if (process.env.NODE_ENV !== 'production' && args[0][0] === undefined) {
        console.error(ILLEGAL_ESCAPE_SEQUENCE_ERROR)
      }

      styles.push(args[0][0])
      const len = args.length
      let i = 1
      for (; i < len; i++) {
        if (process.env.NODE_ENV !== 'production' && args[0][i] === undefined) {
          console.error(ILLEGAL_ESCAPE_SEQUENCE_ERROR)
        }

        styles.push(args[i], args[0][i])
      }
    }

    const Styled = {
      inheritAttrs: false,
      inject: {
        theme: {
          default: undefined
        }
      },

      render(renderContext) {
        const { theme } = renderContext
        const { $attrs, $options, $slots, $props, $parent }= renderContext

        const cache = $parent.$emotionCache
        const { as, ...restAttrs } = $attrs || {}

        let className = ''
        const finalTag = as || baseTag
        const classInterpolations = []
        const mergedProps = {
          ...$attrs,
          theme: theme || $options.inject.theme,
        }

        if ($attrs.class) {
          className += getRegisteredStyles(
            cache.registered,
            classInterpolations,
            $attrs.class
          )
        }

        const serialized = serializeStyles(
          styles.concat(classInterpolations),
          cache.registered,
          mergedProps
        )

        insertStyles(
          cache,
          serialized,
          typeof finalTag === 'string'
        )

        className += `${cache.key}-${serialized.name}`
        if (targetClassName !== undefined) {
          className += ` ${targetClassName}`
        }

        const renderProps = {
            ...$props,
            ...(options.getAttrs ? options.getAttrs(restAttrs) : restAttrs),
            class: className,
        }

        return h(
          finalTag,
          renderProps,
          $slots
        )
      }
    }

    Styled.name =
      identifierName === undefined ?
        `Styled${
          typeof baseTag === 'string' ? baseTag : baseTag.name || 'Component'
        }` :
        identifierName

    Styled.__emotion_real = Styled
    Styled.__emotion_base = baseTag
    Styled.__emotion_styles = styles

    Object.defineProperty(Styled, 'toString', {
      value() {
        if (
          targetClassName === undefined &&
          process.env.NODE_ENV !== 'production'
        ) {
          return 'NO_COMPONENT_SELECTOR'
        }

        return `.${targetClassName}`
      }
    })

    Styled.withComponent = (nextTag, nextOptions) => {
      return createStyled(
        nextTag,
        nextOptions === undefined ?
          options :
          { ...(options || {}), ...nextOptions }
      )(...styles)
    }

    return Styled
  }
}

export const styled = createStyled
