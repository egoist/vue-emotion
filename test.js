import { mount } from '@vue/test-utils'
import styled, { css } from './src'

const props = {
  display: {
    type: String,
    default: 'block'
  },
  padding: {
    type: String,
    default: '10px'
  },
  valid: {
    type: Boolean,
    default: false
  }
}

const displayStyles = ({ display }) => css`
  display: ${display};
`

const paddingStyles = ({ padding }) => css`
  padding: ${padding};
`

const StyledBox = styled('div', { props })`
  text-align: center;
  ${displayStyles};
  ${paddingStyles};
  color: ${props => (props.valid ? 'green' : 'red')};
  background-color: ${props => (props.theme === 'light' ? 'white' : 'blue')};
`

const Box = {
  name: 'Box',
  render() {
    return (
      <StyledBox {...{ props: this.$props, attrs: this.$attrs }}>
        Hello
      </StyledBox>
    )
  }
}

test('it should apply styles to a component with custom classes and props', () => {
  const wrapper = mount(Box, {
    attrs: {
      display: 'flex',
      as: 'main'
    }
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('"as" attribute should work as expected', () => {
  const wrapper = mount(Box, {
    attrs: {
      as: 'section'
    }
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('withComponent should work as expected', () => {
  const wrapper = mount(StyledBox.withComponent('aside'))
  expect(wrapper.element).toMatchSnapshot()
})

test('theme injection should work as expected', () => {
  const wrapper = mount(Box, {
    provide: {
      theme: 'light'
    }
  })

  expect(wrapper.element).toMatchSnapshot()
})

test('can set the theme via an attribute', () => {
  const wrapper = mount(Box, {
    attrs: {
      theme: 'light'
    }
  })

  expect(wrapper.element).toMatchSnapshot()
})
