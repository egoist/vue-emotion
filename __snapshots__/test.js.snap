// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"as" attribute should work as expected 1`] = `
.emotion-0 {
  text-align: center;
  display: block;
  padding: 10px;
  color: red;
  background-color: blue;
}

<div
  as="section"
  class="emotion-0"
>
  Hello
</div>
`;

exports[`can set the theme via an attribute 1`] = `
.emotion-0 {
  text-align: center;
  display: block;
  padding: 10px;
  color: red;
  background-color: white;
}

<div
  class="emotion-0"
  theme="light"
>
  Hello
</div>
`;

exports[`it should apply styles to a component with custom classes and props 1`] = `
.emotion-0 {
  text-align: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 10px;
  color: red;
  background-color: blue;
}

<div
  as="main"
  class="emotion-0"
>
  Hello
</div>
`;

exports[`theme injection should work as expected 1`] = `
.emotion-0 {
  text-align: center;
  display: block;
  padding: 10px;
  color: red;
  background-color: white;
}

<div
  class="emotion-0"
>
  Hello
</div>
`;

exports[`withComponent should work as expected 1`] = `
.emotion-0 {
  text-align: center;
  display: block;
  padding: 10px;
  color: red;
  background-color: blue;
}

<aside
  class="emotion-0 emotion-1"
/>
`;
